### ISSUE TITLE
Scope authorization

### ISSUE THREAD

**Issue by rosshigh** [dwyl/hapi-auth-jwt2#92]
I have the example working and I'm trying to use the scope in the credentials object:
```js
server.route({
    method: 'GET',
    path: '/test',
    handler: function (request, reply) {
        ...
    },
    config: {
        auth: {
            strategy: 'jwt',
            scope: ['admin']
        }
    }
});
```
However, the credentials object doesn't appear to be present so authorization is denied.  Is there a way to use the scope for authorization when using hapi-auth-jwt2?

Thanks,

Ross

**Comment by nelsonic** [comment]
Hi @rosshigh we do not *currently* have support for scopes because it has not been *requested* by anyone *to date*. We do our access control **after** decoding the JWT using custom post-auth logic.
The hapi.js scopes system is good but did not fit our particular needs @DWYL so we wrote a custom one... But, if you would like to ***collaborate*** on adding scopes support to this plugin, we'd be happy to help out where ever we can! :+1:

**Comment by rosshigh** [comment]
Not sure if I'm missing something but I changed the validate function to this:

server.register(require('hapi-auth-jwt2'), function (err) {

    if(err){
        console.log(err);
    }

    server.auth.strategy('jwt', 'jwt', true,
        { key: 'NeverShareYourSecret',          // Never Share your secret key
            verifyOptions: { algorithms: [ 'HS256' ] },
            validateFunc: function (decoded, request, callback) {
                return callback(null, true);
            }
        });
});

And the scope works.  Since decoded is already verified, if all I'm interested in is whether the user is sending a valid token, this will work.

**Comment by nelsonic** [comment]
@rosshigh yes, this will work if all you are interested in is the validity of the token. :+1:

**Comment by nelsonic** [comment]
@rosshigh if this is working for you, please let us know and/or close the issue. thanks! :+1:

**Comment by jamesdixon** [comment]
@nelsonic @rosshigh I looked into this a little further and it doesn't appear anything needs to happen on the `hapi-auth-jwt` side of things. At a basic level, all I did was add a `scope` property to my JWT tokens. From there, in my `validateFunc` function, I simply return my `decoded` token in the callback like so:

```javascript
function (decoded, request, callback) {
// your checks go here    
return callback(null, true, decoded);
}
```

From there, you simply add your required to scope to your route and you're good to go. You obviously do not have to pass back your entire decoded token into the callback and instead, customize it to what you want stored.

If I'm missing something, feel free to let me know.

Cheers,
James

**Comment by nelsonic** [comment]
@jamesdixon thanks for confirming that. :+1:

**Comment by jamesdixon** [comment]
@nelsonic you mentioned that hapi's scopes didn't fit your needs. any chance you could elaborate? i'm in the midst of making some decisions around this and would like to hear what others are doing. thanks!

**Comment by Kitanotori** [comment]
It would be helpful to have support for scopes. Only in the most basic applications all authenticated users are given the same authorizations, so in my opinion lack of support for scopes and lack of scopes in examples is a major problem.

**Comment by jamesdixon** [comment]
@Kitanotori what do you mean by "explicit support for scopes?" From my experience, scopes actually live outside the realm of authentication. See my previous comment that contains the code sample.

**Comment by mcorbe** [comment]
@Kitanotori , I struggled with the same question. The initial example above does not seem to be up to date with version 16.1 (latest) of hapijs hence the confusion, at least for me. The reference here : https://hapijs.com/api/16.1.1#route-options.

TL;DR; Correct path to scope is config.auth.access.scope -> ['admin']
```
server.route({
    method: 'GET',
    path: '/test',
    handler: function (request, reply) {
        ...
    },
    config: {
        auth: {
            strategy: 'jwt',
            access: [{
              scope: ['admin']
            }]
        }
    }
});
```

@jamesdixon please correct me if I am wrong.
Entities and scopes are validated in hapijs core library (https://github.com/hapijs/hapi/blob/master/lib/auth.js#L433) that is the token identification acts additionally on each request and gives you a chance to act JWT authorization. However the scope logic from hapijs still works independently of using this hapi-auth-jwt2 plugin.

**Comment by farzadso** [comment]
@rosshigh @mcorbe 
Thanks guys. I used a mixture of your comments and came up with this solution for my own project. Just thought I'd share it with everyone else.

This is how I registered the `hapi-auth-jwt2` plugin :

```javascript
const hapiAuthJWT = require('hapi-auth-jwt2');

// Hapi JWT2 Auth Plugin
await server.register(hapiAuthJWT);

server.auth.strategy('jwt', 'jwt', {
    key: MY_SUPER_SECRET_KEY,
    validate,
    verifyOptions: {
        algorithms: ['HS256']
    },
});
```

This is my validate function (note: I'm using redis to store my sessions, so don't pay attention to the redis part) :

```javascript
const validate = async function (decoded, request, h) {
    // do your checks to see if the session is valid
    return getAsync(decoded.id).then((reply, error) => {
        if (error) {
            log(error);
        }
        let session;
        if (reply) {
            session = JSON.parse(reply);
        } else { 
            // unable to find session in redis ... reply is null
            return { isValid: false };
        }
        if (session.valid) {
            return { isValid: true };
        } else {
            return { isValid: false };
        }
    });
};
```


I generate a different token for admin users by creating such a token (adding the `admin` scope to the token):

```javascript
let session = {
    valid: true, // this will be set to false when the person logs out
    id: aguid(), // a random session id
    exp: new Date().getTime() + 30 * 24 * 60 * 60 * 1000, // expires in 30 minutes time
    user_id: _user._id,
    scope: 'admin',
};
```
Then I sign it with my secret key :

```javascript
const JWT = require('jsonwebtoken');
...
...
const token = JWT.sign(session, MY_SUPER_SECRET_KEY); 
```

And include this token in the response header :

```javascript
const loginHandler = (request, h) => {
    h.header('Authorization', token);
    return h;
}
```
This is the configuration I use for my path :

```javascript
 {
     path: '/super-secret',
     method: 'POST',
     config: {
         auth: {
             strategy: 'jwt',
             access: [{
                 scope: ['admin']
             }]
         },
     },
     handler: superSecretRouteHandler,
 }
```

**Comment by michalczukm** [comment]
@farzadso I had the same issue - you have to return scopes in `credentials` field. Check mine code below

```
type DecodedAuth0JwtToken = {
    sub: string,
    aud: string[],
    iat: number,
    exp: number,
    azp: string,
    scope?: string
};

const validateUser = async (decoded: DecodedAuth0JwtToken | null) => {
    if (decoded && decoded.sub) {
        return decoded.scope
            ? {
                isValid: true,
                credentials: {
                    scope: decoded.scope.split(' ')
                }
            }
            : { isValid: true };
    }

    return { isValid: false };
};
```

and the strategy (for context)
```
serverInstance.auth.strategy('jwt', 'jwt', {
        complete: true,
        key: jwksRsa.hapiJwt2KeyAsync({
            cache: true,
            rateLimit: true,
            jwksRequestsPerMinute: 20,
            jwksUri: env.auth.jwksUri
        }),
        verifyOptions: {
            audience: '/admin',
            issuer: env.auth.issuer,
            algorithms: ['RS256']
        },
        validate: validateUser
    });
```

**Comment by michalczukm** [comment]
@farzadso the `credentials` object is handling all informations required by auth module/plugin of hapi. Probably hapi > 17 or hapi-auth-jwt2 did it automatically, but not anymore :)

PS it works for valid tokens in your case, right?

**Comment by farzadso** [comment]
@michalczukm I'm on the latest versions on hapi and hapi-auth-jwt2 and it's still working. 

yes, whenever I add this to the config of a path :

```
config: {
         auth: {
             strategy: 'jwt',
             access: [{
                 scope: ['admin']
             }]
         },
     },
```

my token must have the admin scope inside it

**Comment by michalczukm** [comment]
@farzadso so its working correctly :)
